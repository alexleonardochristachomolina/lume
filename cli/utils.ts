import lume from "../mod.ts";
import { toFileUrl } from "../deps/path.ts";
import { isUrl } from "../core/utils/path.ts";
import { getConfigFile } from "../core/utils/lume_config.ts";
import { log } from "../core/utils/log.ts";
import Site from "../core/site.ts";

/** Create a site instance */
export async function createSite(config?: string): Promise<Site> {
  let url: string | undefined;
  const watchInclude = [];

  if (config && isUrl(config)) {
    url = config;
  } else {
    const path = await getConfigFile(config);

    if (path) {
      watchInclude.push(path);
      url = toFileUrl(path).href;
    }
  }

  if (url) {
    log.info(`Loading config file <gray>${url}</gray>`);
    const mod = await import(url);
    const site = mod.default as Site | undefined;
    if (!(site instanceof Site)) {
      log.critical(
        `[Lume] Missing Site instance! Ensure your config file does export the Site instance as default.`,
      );
      throw new Error("Site instance is not found");
    }

    site._data.configFile = url;
    site.options.watcher.include.push(...watchInclude);
    return site;
  }

  return lume();
}

/** Create a site intance and build it */
export async function buildSite(config?: string): Promise<Site> {
  const site = await createSite(config);

  performance.mark("start");
  await site.build();
  performance.mark("end");

  log.info(`üçæ Site built into <gray>${site.options.dest}</gray>`);
  const duration = performance.measure("duration", "start", "end").duration /
    1000;
  const total = site.pages.length + site.files.length;
  log.info(
    `  <gray>${total} files generated in ${duration.toFixed(2)} seconds</gray>`,
  );

  return site;
}
