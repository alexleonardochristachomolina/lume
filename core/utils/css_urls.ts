type UrlWalker = (url: string, type: string) => string | Promise<string>;
type Position = [string, string, number, number];
type Index = [string, number];

export async function walkUrls(
  code: string,
  walker: UrlWalker,
): Promise<string> {
  let result = "";
  let index = 0;
  let position = next(code, index);

  while (position) {
    const [type, url, start, end] = position;
    const newUrl = await walker(url, type);
    result += `${code.slice(index, start)}"${newUrl}"`;
    index = end;
    position = next(code, index);
  }

  return result + code.slice(index);
}

function next(code: string, index = 0): Position | undefined {
  const nextUrl = code.indexOf("url(", index);
  const nextComment = code.indexOf("/*", index);
  const nextImport = code.indexOf("@import ", index);

  let found: Index | undefined;

  if (nextUrl !== -1) {
    found = ["url", nextUrl];
  }
  if (nextComment !== -1 && (!found || nextComment < found[1])) {
    found = ["comment", nextComment];
  }
  if (nextImport !== -1 && (!found || nextImport < found[1])) {
    found = ["import", nextImport];
  }

  if (!found) {
    return;
  }

  // If it's a comment, find the end of it
  if (found[0] === "comment") {
    const end = code.indexOf("*/", found[1]);
    if (end === -1) {
      return;
    }
    return next(code, end + 2);
  }

  if (found[0] === "import") {
    const double = code.indexOf('"', found[1]);
    const single = code.indexOf("'", found[1]);
    const urlFn = code.indexOf("url(", found[1]);

    let foundImport: Index | undefined;

    if (double !== -1) {
      foundImport = ["double", double];
    }
    if (single !== -1 && (!foundImport || single < foundImport[1])) {
      foundImport = ["single", single];
    }

    // It's has a url(), so delegate to the url() logic
    if (urlFn !== -1 && (!foundImport || urlFn < foundImport[1])) {
      return next(code, urlFn);
    }

    if (!foundImport) {
      return;
    }

    const start = foundImport[1];
    const end =
      code.indexOf(foundImport[0] === "double" ? '"' : "'", start + 1) + 1;
    return ["url", code.slice(start + 1, end - 1), start, end];
  }

  if (found[0] === "url") {
    const start = code.indexOf("(", found[1]) + 1;
    const end = code.indexOf(")", start);
    let url = code.slice(start, end).trim();
    // Remove quotes
    if (url.startsWith('"') || url.startsWith("'")) {
      url = url.slice(1, -1);
    }
    return ["url", url, start, end];
  }
}
