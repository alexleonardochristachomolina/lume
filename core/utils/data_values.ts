import type { Data } from "../file.ts";

/**
 * Get the value of a page data
 * For example, if the value is "=title", it returns the value of the page data "title"
 * If the value is "$.title", it will return the value of the element with the selector ".title"
 */
export function getDataValue(data: Partial<Data>, value?: unknown) {
  // Get the value from the page data
  if (typeof value === "string") {
    return searchValue(data, value);
  }

  if (typeof value === "function") {
    return value(data);
  }

  return value;
}

function searchValue(data: Partial<Data>, value: string): unknown {
  if (!value) {
    return;
  }

  if (value.startsWith("=")) {
    let key = value.slice(1);
    [key, value] = parseFallback(key);

    if (!key.includes(".")) {
      return data[key] ?? searchValue(data, value);
    }

    const keys = key.split(".");
    let val = data;
    for (const key of keys) {
      val = val?.[key];
    }
    return val ?? searchValue(data, value);
  }

  if (value.startsWith("$")) {
    let selector = value.slice(1);
    [selector, value] = parseFallback(selector);

    return queryCss(selector, data.page?.document) ?? searchValue(data, value);
  }

  return value;
}

function parseFallback(key: string): [string, string] {
  const fallback = key.indexOf("||");

  if (fallback !== -1) {
    return [
      key.slice(0, fallback).trim(),
      key.slice(fallback + 2).trim(),
    ];
  }

  return [
    key,
    "",
  ];
}

// https://regexr.com/7qnot
const checkForAttrPattern = /^(.+)\s+(?:attr\(([\w\-]+)\))$/;

function queryCss(query: string, document?: Document) {
  const checkResult = query.match(checkForAttrPattern);

  const hasAttr = checkResult?.[0];
  if (hasAttr) {
    const [_, query, name] = checkResult;
    return document?.querySelector(query)?.getAttribute(name);
  }

  return document?.querySelector(query)?.innerHTML;
}
