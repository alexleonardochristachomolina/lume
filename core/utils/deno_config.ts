import { parse } from "../../deps/jsonc.ts";
import { isUrl } from "../utils/path.ts";

/** Import map file */
export interface ImportMap {
  imports: Record<string, string>;
  scopes?: Record<string, Record<string, string>>;
}

/** Basic options for deno.json file */
export interface DenoConfig extends Partial<ImportMap> {
  importMap?: string;
  tasks?: Record<string, string>;
  compilerOptions?: {
    jsx?: "jsx" | "react-jsx" | "precompile";
    jsxImportSource?: string;
    types?: string[];
  };
  [key: string]: unknown;
}

export interface DenoConfigResult {
  file: string;
  config: DenoConfig;
  importMap?: ImportMap;
}

/** Detect and returns the Deno configuration */
export async function readDenoConfig(): Promise<DenoConfigResult | undefined> {
  for (const file of ["deno.json", "deno.jsonc"]) {
    try {
      const content = await Deno.readTextFile(file);
      const config = parse(content) as DenoConfig;
      let importMap: ImportMap | undefined;

      if (config.importMap) {
        importMap = isUrl(config.importMap)
          ? await (await fetch(config.importMap)).json()
          : await JSON.parse(await Deno.readTextFile(config.importMap));
      } else if (config.imports) {
        importMap = {
          imports: config.imports,
          scopes: config.scopes,
        };
      }
      return { file, config, importMap };
    } catch (err) {
      if (err instanceof Deno.errors.NotFound) {
        continue;
      }

      throw err;
    }
  }
}
