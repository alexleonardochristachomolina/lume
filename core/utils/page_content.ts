import type { Page } from "../file.ts";

export function insertContent(
  page: Page,
  content: string,
  placeholder?: string,
) {
  const pageContent = page.content;

  if (pageContent && typeof pageContent === "string") {
    if (placeholder && pageContent.includes(placeholder)) {
      page.content = pageContent.replace(
        placeholder,
        content,
      );
    } else {
      page.content += `\n${content}`;
    }
  } else {
    page.content = content;
  }
}
