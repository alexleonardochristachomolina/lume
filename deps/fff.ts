export type * from "https://deno.land/x/fff@v1.2.1/src/types.ts";

export {
  postTypeDiscovery,
} from "https://deno.land/x/fff@v1.2.1/src/utils/ptd.ts";

export {
  type FFFTransformPreset,
  transform,
} from "https://deno.land/x/fff@v1.2.1/src/utils/transform.ts";

export {
  strict,
  type StrictPresetOptions,
} from "https://deno.land/x/fff@v1.2.1/src/utils/presets/strict.ts";
