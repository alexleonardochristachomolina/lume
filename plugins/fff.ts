import type Site from "../core/site.ts";
import { getGitDate } from "../core/utils/date.ts";
import { merge } from "../core/utils/object.ts";

import {
  type FFFFlavoredFrontmatter,
  type FFFTransformPreset,
  postTypeDiscovery,
  strict,
  type StrictPresetOptions,
  transform,
} from "../deps/fff.ts";

export interface Options {
  /** The list extensions this plugin applies to */
  extensions: string[];
  /**
   * The list of presets to apply
   * @see https://fff.js.org/concepts/flavor-transform.html#fff-transform-preset
   */
  presets: FFFTransformPreset[];
  strict: false | StrictPresetOptions;
  /** To convert the generic `date` field to one of these values */
  date?: "created" | "updated" | "published";
  /** Get the date from the git history */
  getGitDate?: boolean;
  postTypeDiscovery?: boolean;
}

// Default options
export const defaults: Options = {
  extensions: [".html"],
  presets: [],
  strict: false,
};

/**
 * A plugin to transform frontmatter using FFF
 * @see https://lume.land/plugins/fff/
 */
export function fff(userOptions?: Partial<Options>) {
  const options = merge(defaults, userOptions);

  return (site: Site) => {
    site.preprocess(options.extensions, (pages) =>
      pages.forEach((page) => {
        if (options.getGitDate && page.src.entry) {
          if (!page.data.created) {
            page.data.created = getGitDate("created", page.src.entry.src);
          }
          if (!page.data.updated) {
            page.data.updated = getGitDate("modified", page.src.entry.src);
          }
        }

        page.data = transform(page.data, [
          ...options.presets,
          ...(options.date ? [{ "date": options.date }] : []),
          ...(options.strict === false ? [] : [strict(options.strict)]),
        ]);

        if (options.postTypeDiscovery) {
          page.data.type = postTypeDiscovery(page.data);
        }
      }));
  };
}

export default fff;

/** Extends Data interface */
declare global {
  namespace Lume {
    export interface Data
      extends Omit<FFFFlavoredFrontmatter, "lang" | "tags"> {}
  }
}
