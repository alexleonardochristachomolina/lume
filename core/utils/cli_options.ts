import { parseArgs } from "../../deps/cli.ts";
import type { DeepPartial } from "./object.ts";
import type { SiteOptions } from "../site.ts";

export function getOptionsFromCli(
  options: DeepPartial<SiteOptions>,
): DeepPartial<SiteOptions> {
  const cli = parseArgs(Deno.args, {
    string: ["src", "dest", "location", "port"],
    boolean: ["serve", "open"],
    alias: { dev: "d", serve: "s", port: "p", open: "o" },
    ["--"]: true,
  });

  if (cli.src) {
    options.src = cli.src;
  }

  if (cli.dest) {
    options.dest = cli.dest;
  }

  const serveMode = cli.serve || cli._[0] === "cms";

  if (cli.port) {
    (options.server ||= {}).port = parseInt(cli.port);
  } else if (serveMode) {
    (options.server ||= {}).port = 3000;
  }

  let location: URL;

  if (cli.location) {
    location = new URL(cli.location);
  } else if (options.location && !serveMode) {
    location = options.location as URL;
  } else {
    location = new URL("http://localhost");
  }

  let port: number;

  if (cli.port) {
    port = parseInt(cli.port);
  } else if (location.port) {
    port = parseInt(location.port);
  } else if (options.server?.port) {
    port = options.server.port;
  } else {
    port = cli.serve ? 3000 : location.protocol === "https:" ? 443 : 80;
  }

  (options.server ||= {}).port = port;
  location.port = port.toString();
  options.location = location;

  if (cli.open) {
    (options.server ||= {}).open = cli.open;
  }

  return options;
}
