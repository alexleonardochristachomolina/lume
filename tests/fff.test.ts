import { assertEquals } from "../deps/assert.ts";
import { assertSiteSnapshot, build, getSite } from "./utils.ts";
import fff from "../plugins/fff.ts";

Deno.test("FFF plugin", async (t) => {
  const site = getSite({
    src: "fff",
    location: new URL("https://example.com/"),
  });

  site.use(fff({
    date: "published",
    presets: [{
      summary: "content",
    }],
    strict: {
      categories: false,
      media: {
        array: false,
        type: "object",
      },
    },
    getGitDate: true,
    postTypeDiscovery: true,
  }));

  // build page
  await build(site);

  // compare to snapshot
  await assertSiteSnapshot(t, site);

  const { pages } = site;

  // published => date
  const date = pages.find((page) => page.src.path === "/date")!;
  assertEquals(date.data.date, date.data.published);
  // getGitDate
  assertEquals(!!date.data.created, true);
  assertEquals(!!date.data.updated, true);
  // postTypeDiscovery
  assertEquals(date.data.type, "article");

  // images (string media) => image (object media)
  const image = pages.find((page) => page.src.path === "/image")!;
  assertEquals(image.data.image.alt, "FFF Image Test");
  assertEquals(image.data.image.src, "/my-image.png");
  // getGitDate
  assertEquals(!!image.data.created, true);
  assertEquals(!!image.data.updated, true);
  // postTypeDiscovery
  assertEquals(image.data.type, "photo");

  // categories => tags
  const tags = pages.find((page) => page.src.path === "/tags")!;
  assertEquals(tags.data.tags, ["foo", "bar", "baz", "qux"]);
  // getGitDate
  assertEquals(!!tags.data.created, true);
  assertEquals(!!tags.data.updated, true);
  // postTypeDiscovery (todo: fix this)
  // assertEquals(tags.data.type, "note");
});
