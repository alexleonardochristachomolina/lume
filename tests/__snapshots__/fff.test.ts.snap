export const snapshot = {};

snapshot[`FFF plugin 1`] = `
{
  formats: [
    {
      engines: 0,
      ext: ".page.toml",
      loader: [AsyncFunction: toml],
      pageType: "page",
    },
    {
      engines: 1,
      ext: ".page.ts",
      loader: [AsyncFunction: module],
      pageType: "page",
    },
    {
      engines: 1,
      ext: ".page.js",
      loader: [AsyncFunction: module],
      pageType: "page",
    },
    {
      engines: 0,
      ext: ".page.jsonc",
      loader: [AsyncFunction: json],
      pageType: "page",
    },
    {
      engines: 0,
      ext: ".page.json",
      loader: [AsyncFunction: json],
      pageType: "page",
    },
    {
      dataLoader: [AsyncFunction: json],
      engines: 0,
      ext: ".json",
      loader: [AsyncFunction: json],
    },
    {
      dataLoader: [AsyncFunction: json],
      engines: 0,
      ext: ".jsonc",
      loader: [AsyncFunction: json],
    },
    {
      engines: 1,
      ext: ".md",
      loader: [AsyncFunction: text],
      pageType: "page",
    },
    {
      engines: 1,
      ext: ".markdown",
      loader: [AsyncFunction: text],
      pageType: "page",
    },
    {
      dataLoader: [AsyncFunction: module],
      engines: 1,
      ext: ".js",
      loader: [AsyncFunction: module],
    },
    {
      dataLoader: [AsyncFunction: module],
      engines: 1,
      ext: ".ts",
      loader: [AsyncFunction: module],
    },
    {
      engines: 1,
      ext: ".vento",
      loader: [AsyncFunction: text],
      pageType: "page",
    },
    {
      engines: 1,
      ext: ".vto",
      loader: [AsyncFunction: text],
      pageType: "page",
    },
    {
      dataLoader: [AsyncFunction: toml],
      engines: 0,
      ext: ".toml",
      loader: [AsyncFunction: toml],
    },
    {
      dataLoader: [AsyncFunction: yaml],
      engines: 0,
      ext: ".yaml",
      loader: [AsyncFunction: yaml],
      pageType: "page",
    },
    {
      dataLoader: [AsyncFunction: yaml],
      engines: 0,
      ext: ".yml",
      loader: [AsyncFunction: yaml],
      pageType: "page",
    },
  ],
  src: [
    "/",
    "/date.md",
    "/image.md",
    "/tags.md",
  ],
}
`;

snapshot[`FFF plugin 2`] = `[]`;

snapshot[`FFF plugin 3`] = `
[
  {
    content: "<!DOCTYPE html>
<p>The <code>published</code> should be used as <code>date</code>.</p>
",
    data: {
      alt: "undefined",
      audio: "undefined",
      basename: "date",
      categories: "undefined",
      children: "<p>The <code>published</code> should be used as <code>date</code>.</p>
",
      content: "The \`published\` should be used as \`date\`.
",
      created: [],
      date: [],
      draft: "undefined",
      flags: "undefined",
      image: "undefined",
      images: "Array(0)",
      mergedKeys: [
        "tags",
      ],
      page: [
        "src",
        "data",
        "asset",
      ],
      paginate: "paginate",
      published: [],
      search: [],
      summary: "The \`published\` should be used as \`date\`.
",
      tags: "Array(0)",
      title: "FFF Date Test",
      type: "article",
      updated: [],
      url: "/date/",
      video: "undefined",
      visibility: "undefined",
    },
    src: {
      asset: false,
      ext: ".md",
      path: "/date",
      remote: undefined,
    },
  },
  {
    content: "<!DOCTYPE html>
<p>FFF Image Test</p>
",
    data: {
      alt: "FFF Image Test",
      audio: "undefined",
      basename: "image",
      categories: "undefined",
      children: "<p>FFF Image Test</p>
",
      content: "FFF Image Test
",
      created: [],
      date: [],
      draft: "undefined",
      flags: "undefined",
      image: [
        "alt",
        "src",
      ],
      images: "Array(1)",
      mergedKeys: [
        "tags",
      ],
      page: [
        "src",
        "data",
        "asset",
      ],
      paginate: "paginate",
      search: [],
      summary: "FFF Image Test
",
      tags: "Array(0)",
      type: "photo",
      updated: [],
      url: "/image/",
      video: "undefined",
      visibility: "undefined",
    },
    src: {
      asset: false,
      ext: ".md",
      path: "/image",
      remote: undefined,
    },
  },
  {
    content: "<!DOCTYPE html>
<p>FFF Tags Test</p>
<p>The <code>categories</code> should be merged into <code>tags</code>.</p>
",
    data: {
      alt: "undefined",
      audio: "undefined",
      basename: "tags",
      categories: "Array(2)",
      children: "<p>FFF Tags Test</p>
<p>The <code>categories</code> should be merged into <code>tags</code>.</p>
",
      content: "FFF Tags Test

The \`categories\` should be merged into \`tags\`.
",
      created: [],
      date: [],
      draft: "undefined",
      flags: "undefined",
      image: "undefined",
      images: "Array(0)",
      mergedKeys: [
        "tags",
      ],
      page: [
        "src",
        "data",
        "asset",
      ],
      paginate: "paginate",
      search: [],
      summary: "FFF Tags Test

The \`categories\` should be merged into \`tags\`.
",
      tags: "Array(4)",
      type: "photo",
      updated: [],
      url: "/tags/",
      video: "undefined",
      visibility: "undefined",
    },
    src: {
      asset: false,
      ext: ".md",
      path: "/tags",
      remote: undefined,
    },
  },
]
`;
